src/ratelimit.rs

```diff
@@ -0,0 +1,137 @@
+//! # Rate limiting module.
+//!
+//! This module contains implementation of token bucket policy.
+//! Its primary use is preventing Delta Chat from sending too many messages, especially automatic,
+//! such as read receipts.
+
+use std::time::{Duration, SystemTime};
+
+#[derive(Debug)]
+pub(crate) struct Ratelimit {
+    /// Time of the last update.
+    last_update: SystemTime,
+
+    /// Number of messages sent within the time window ending at `last_update`.
+    current_value: f64,
+
+    /// Time window size.
+    window: Duration,
+
+    /// Number of messages allowed to send within the time window.
+    quota: f64,
+}
+
+impl Ratelimit {
+    /// Returns a new rate limiter with the given constraints.
+    ///
+    /// Rate limiter will allow to send no more than `quota` messages within duration `window`.
+    pub(crate) fn new(window: Duration, quota: f64) -> Self {
+        Self::new_at(window, quota, SystemTime::now())
+    }
+
+    /// Returns a new rate limiter with given current time for testing purposes.
+    const fn new_at(window: Duration, quota: f64, now: SystemTime) -> Self {
+        Self {
+            last_update: now,
+            current_value: 0.0,
+            window,
+            quota,
+        }
+    }
+
+    /// Update current value.
+    pub(crate) fn update_at(&mut self, now: SystemTime) {
+        let rate: f64 = self.quota / self.window.as_secs_f64();
+        let elapsed = now
+            .duration_since(self.last_update)
+            .unwrap_or(Duration::ZERO)
+            .as_secs_f64()
+            .max(0.0);
+        self.current_value = (self.current_value - rate * elapsed).max(0.0);
+        self.last_update = now;
+    }
+
+    /// Returns true if it is allowed to send a message.
+    fn can_send_at(&mut self, now: SystemTime) -> bool {
+        self.update_at(now);
+        self.current_value <= self.quota
+    }
+
+    /// Returns true if can send another message now.
+    pub(crate) fn can_send(&mut self) -> bool {
```

Comment:
i find it it bit unexpected, that the function checking for send-ability modifies the state (and needs a &mut therefore).

probably somehow needed by the algorighm, maybe add a line and explain why this is needed?